sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,370 @@
+	SubjectUid uint32
+
+	// Path is the path of the file, as seen by the process triggering the request.
+	Path string
+	// Permission is the opaque permission that is being requested.
+	Permission interface{}
+	// YesNo is a channel for writing the response.
+	YesNo chan bool
+}
+
+func newRequest(msg *notify.MsgNotificationFile) (*Request, error) {
+	var perm interface{}
+	if msg.Class == notify.AA_CLASS_FILE {
+		_, missingPerms, err := msg.DecodeFilePermissions()
+		if err != nil {
+			return nil, err
+		}
+		perm = missingPerms
+	}
+	return &Request{
+		Pid:        msg.Pid,
+		Label:      msg.Label,
+		Path:       msg.Name,
+		SubjectUid: msg.SUID,
+
+		Permission: perm,
+
+		YesNo: make(chan bool, 1),
+	}, nil
+}
+
+// Listener contains low-level components for receiving notification requests
+// and responding with notification responses.
+type Listener struct {
+	// reqs is a channel with incoming requests. Each request is asynchronous
+	// and needs to be replied to.
+	reqs chan *Request
+	// errs is a channel for receiving asynchronous error messages from
+	// concurrently running parts of the listener system.
+	errs chan error
+
+	notifyFile *os.File
+	poll       *epoll.Epoll
+
+	tomb tomb.Tomb
+
+	// status keeps track of whether the listener has been run and/or closed,
+	// both to ensure that Run() and Close() are executed at most once each,
+	// and to ensure that the listener cannot be run after it has been closed.
+	// Must be read/modified atomically, and can only be changed in one of the
+	// following ways:
+	// - statusReady -> statusRunning
+	// - statusReady -> statusClosed
+	// - statusRunning -> statusClosed
+	status uint32
+}
+
+const (
+	statusReady uint32 = iota
+	statusRunning
+	statusClosed
+)
+
+// Register opens and configures the apparmor notification interface.
+//
+// If the kernel does not support the notification mechanism the error is ErrNotSupported.
+func Register() (*Listener, error) {
+	path := notify.SysPath
+	if override := os.Getenv("PROMPT_NOTIFY_PATH"); override != "" {
+		path = override
+	}
+
+	notifyFile, err := osOpen(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrListenerNotSupported
+		}
+		return nil, fmt.Errorf("cannot open %q: %v", path, err)
+	}
+	defer func() {
+		if err != nil {
+			notifyFile.Close()
+		}
+	}()
+
+	msg := notify.MsgNotificationFilter{ModeSet: notify.APPARMOR_MODESET_USER}
+	data, err := msg.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	ioctlBuf := notify.IoctlRequestBuffer(data)
+	_, err = notifyIoctl(notifyFile.Fd(), notify.APPARMOR_NOTIF_SET_FILTER, ioctlBuf)
+	if err != nil {
+		return nil, fmt.Errorf("cannot notify ioctl to modeset user on %q: %v", path, err)
+	}
+
+	poll, err := epoll.Open()
+	if err != nil {
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			poll.Close()
+		}
+	}()
+	// XXX: Do we need a notification for Writable, to send responses back?
```

Comment:
I verified this with JJ, there's no need of checking for epoll writable. As long as the kernel is waiting on a request with the given ID, we can write whenever we'd like.